src/main/java/com/uid2/admin/vertx/AdminVerticle.java

```diff
@@ -25,149 +27,115 @@ public class AdminVerticle extends AbstractVerticle {
+        router.route().handler(StaticHandler.create("webroot"));
 
-        router.get("/ops/healthcheck").handler(this::handleHealthCheck);
+        router.route("/login").handler(authHandler);
+        router.route("/adm/*").handler(authHandler);
 
-        router.get("/api/token/get").handler(this::handleTokenGet);
+        router.get("/login").handler(redirectToRootHandler);
+        router.get("/logout").handler(redirectToRootHandler);
+        router.get("/ops/healthcheck").handler(this::handleHealthCheck);
+        router.get("/api/token/get").handler(ctx -> handleTokenGet(ctx, oauth2Provider));
 
         for (IService service : this.services) {
             service.setupRoutes(router);
         }
 
-        router.route().handler(StaticHandler.create("webroot"));
         return router;
     }
 
     private void handleHealthCheck(RoutingContext rc) {
         rc.response().end("OK");
     }
 
-    private void handleTokenGet(RoutingContext rc) {
+    private void handleTokenGet(RoutingContext rc, OAuth2Auth oauth2Provider) {
         if (isAuthDisabled(config)) {
             respondWithTestAdminUser(rc);
-        }
-        else {
-            respondWithRealUser(rc);
+        } else {
+            respondWithRealUser(rc, oauth2Provider);
         }
     }
 
-    private void respondWithRealUser(RoutingContext rc) {
-        AccessToken user = (AccessToken) rc.user();
-        // retrieve the user profile, this is a common feature but not from the official OAuth2 spec
-        user.userInfo(res -> {
-            if (res.failed()) {
-                rc.session().destroy();
-                rc.fail(res.cause());
-            } else {
-                final JsonObject userInfo = res.result();
-                String contact = userInfo.getString("email");
-                if (contact == null) {
-                    user.fetch("https://api.github.com/user/emails", rcEmail -> {
-                        if (rcEmail.failed()) {
-                            rc.session().destroy();
-                            rc.fail(res.cause());
-                        } else {
-                            JsonArray emails = rcEmail.result().jsonArray();
-                            if (emails.size() > 0) {
-                                final String publicEmail = emails.getJsonObject(0).getString("email");
-                                handleEmailContactInfo(publicEmail, rc);
-                            } else {
-                                rc.fail(new Throwable("No public emails"));
-                            }
-                        }
-                    });
-                } else {
-                    handleEmailContactInfo(contact, rc);
-                }
-            }
-        });
+    private void respondWithRealUser(RoutingContext rc, OAuth2Auth oauth2Provider) {
+        oauth2Provider.userInfo(rc.user())
+                .onFailure(e -> {
+                    rc.session().destroy();
+                    rc.fail(e);
```

Comment:
This will return the full error, not just the cause as previously. This could leak information

Reply:
ok